streaming/index.js

```diff
@@ -182,14 +181,69 @@ const CHANNEL_NAMES = [
 ];
 
 const startServer = async () => {
+  const pgPool = new pg.Pool(pgConfigFromEnv(process.env));
+  const server = http.createServer();
+  const wss = new WebSocket.Server({ noServer: true });
+
+  // Set the X-Request-Id header on WebSockets:
+  wss.on("headers", function onHeaders(headers, req) {
+    headers.push(`X-Request-Id: ${req.id}`);
+  });
+
   const app = express();
 
   app.set('trust proxy', process.env.TRUSTED_PROXY_IP ? process.env.TRUSTED_PROXY_IP.split(/(?:\s*,\s*|\s+)/) : 'loopback,uniquelocal');
 
-  const pgPool = new pg.Pool(pgConfigFromEnv(process.env));
-  const server = http.createServer(app);
   app.use(cors());
 
+  // Handle eventsource & other http requests:
+  server.on('request', app);
+
+  // Handle upgrade requests:
+  server.on('upgrade', function upgrade(request, socket, head) {
+    /** @param {Error} err */
+    const onSocketError = (err) => {
+      log.error(`Error with websocket upgrade: ${err}`);
+    };
+
+    socket.on('error', onSocketError);
+
+    // Authenticate:
+    accountFromRequest(request).then(( /* resolvedAccount */) => {
+      wss.handleUpgrade(request, socket, head, function done(ws) {
+        // Remove the error handler:
+        socket.removeListener('error', onSocketError);
+
+        // Start the connection:
+        wss.emit('connection', ws, request);
+      });
+    }).catch((err) => {
+      // Unfortunately for using the on('upgrade') setup, we need to manually
+      // write a HTTP Response to the Socket to close the connection upgrade
+      // attempt, so the following code is to handle all of that.
+      const statusCode = err.status ?? 401;
+
+      // Ensure the socket is closed once we've finished writing to it:
+      socket.once('finish', () => {
+        socket.destroy();
+      });
+
+      /** @type {Record<string, string | number>} */
+      const headers = {
+        'Connection': 'close',
+        'Content-Type': 'text/plain',
+        'Content-Length': 0,
+        'X-Request-Id': request.id.toString(),
+        // Send the error message via header so it can be debugged in developer
+        // tools:
+        'X-Error-Message': err.message
```

Comment:
This may leak internal information about errors to the client, potentially, but with the logging changes, we can capture this error and log to the error log instead, with the correlation ID. We likely need to limit this to errors with a status only, but I'm not sure if all code paths that result in this being reached create the errors appropriately.

There's more refactoring needed around the promises used to fetch the account from the request and valid the initial channel join request (from the URL), currently this is done in the `on("connection")` handler, when it'd be better to terminate the connection before even initialising the WebSocket.

Reply:
I think I'd just leave that out for now, we did not previously return them, so this would not be a regression.

Reply:
Yeah, I've removed this out but left a TODO; it was very helpful when debugging the websocket connections in browser for issues (e.g., invalidating the access token on the socket)